### PR TITLE
 Add E2E tests for X.509 auth

### PR DESCRIPTION
**Pull request by mrohera**
Changes here implement E2E tests for downstream devices to authenticate with the Edge using X.509 certificates. Additional noteworthy change is that the EdgeHub enables X.509 auth to ON by default.

### CODE REVIEW

File: smoke/LeafDevice/details/Details.cs [hunk 1]
```diff
@@ -23,32 +24,98 @@ namespace LeafDevice.Details
...
         readonly string deviceId;
-        readonly string certificateFileName;
+        readonly string trustedCACertificateFileName;
         readonly string edgeHostName;
         readonly ServiceClientTransportType serviceClientTransportType;
         readonly EventHubClientTransportType eventHubClientTransportType;
         readonly ITransportSettings[] deviceTransportSettings;
-
+        readonly AuthenticationType authType = AuthenticationType.None;
+        Option<X509Certificate2> clientCertificate = Option.None<X509Certificate2>();
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: Why are these not ro?

File: smoke/LeafDevice/details/Details.cs [hunk 2]
```diff
@@ -64,41 +131,63 @@ public class Details
             }
         }
 
-        protected Task InitializeServerCerts()
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: Nit: might be cleaner to have 2 separate methods - one for server certs and one for client certs.

File: smoke/LeafDevice/details/Process.cs [hunk 3]
```diff
@@ -23,10 +23,10 @@ public static async Task<string[]> RunAsync(string name, string args, Cancellati
             var info = new ProcessStartInfo
             {
                 FileName = name,
-                Arguments = args,
+                Arguments = args
             };
 
-            using (ProcessResults result = await ProcessEx.RunAsync(info, token))
+            using (var result = await ProcessEx.RunAsync(info, token))
```
[the remark this answers is not in the document]
Comment (mrohera) [reply smoke/LeafDevice/details/Process.cs]: Ha! This was not me I can assure you :). Resharper CLI and the Azure IoT rules reformatted the code here and everywhere else. Other funny thing this is after running the CLI the stylecheck still fails so there is some inconsistency here that needs to be addressed.

File: smoke/LeafDevice/LeafDevice.cs [hunk 4]
```diff
@@ -2,18 +2,77 @@
...
+
+        public LeafDevice(
+            string iothubConnectionString,
+            string eventhubCompatibleEndpointWithEntityPath,
+            string deviceId,
+            string trustedCACertificateFileName,
+            string edgeHostName,
+            bool useWebSockets,
+            string clientCertificatePath,
+            string clientCertificateKeyPath,
```
[the remark this answers is not in the document]
Comment (mrohera) [reply smoke/LeafDevice/LeafDevice.cs]: So for thumbprint auth there are two certs involved a primary and secondary which are both needed to setup a device in IoT Hub. Additionally for a thorough test coverage, we ought to connect client using each of these certs one at a time. 

With that being said, clientCertificatePath and clientCertificateKeyPath are the actual creds a client will authenticate with and the thumprints list will contain the certificates only (no keys) which will be used to establish the thumbprints.

For X.509 CA auth, thumbprints just don't come into play and the clientCertificatePath and clientCertificateKeyPath are used and quite straightforward.

File: smoke/LeafDevice/details/Details.cs (hunk 2 above)
Comment (mrohera) [reply smoke/LeafDevice/details/Details.cs]: At the end, they are both simply trusted certificates that need to get installed and both can have different roots. I felt generalizing this simplified things and also in line with our existing terminology.

File: smoke/LeafDevice/details/Process.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (philipktlin) [reply smoke/LeafDevice/details/Process.cs]: when running Resharper CLI, we need to provide --profile as well and command is updated in OneNote.

File: smoke/LeafDevice/details/Details.cs [hunk 5]
```diff
@@ -64,41 +131,63 @@ public class Details
...
+                    CustomCertificateValidator.Create(new List<X509Certificate2> { this.GetTrustedCertificate() }, this.deviceTransportSettings);
+                }
+                else
+                {
+                    InstallTrustedCACerts(new List<X509Certificate2> { this.GetTrustedCertificate() });
+                }
             }
 
-            return Task.CompletedTask;
+            return this.clientCertificateChain.Map(
```
[the remark this answers is not in the document]
Comment (mrohera) [reply smoke/LeafDevice/details/Details.cs]: Fixed.

File: smoke/LeafDevice/Program.cs [hunk 6]
```diff
@@ -66,14 +85,63 @@ async Task<int> OnExecuteAsync()
...
+                        this.DeviceId,
+                        this.TrustedCACertificateFileName,
+                        this.EdgeHostName,
+                        this.UseWebSockets,
+                        this.X509PrimaryCertPath,
+                        this.X509PrimaryKeyPath,
+                        thumbprintCerts);
+                    await testPrimary.RunAsync();
+
+                    var testSeondary = new LeafDevice(
```
[the remark this answers is not in the document]
Comment (mrohera) [reply smoke/LeafDevice/Program.cs]: Fixed. It was a copy pasta bug. Thanks!

File: smoke/LeafDevice/LeafDevice.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (varunpuranik) [reply smoke/LeafDevice/LeafDevice.cs]: I understand the intent.. seems cleaner to have one cert and first set up Auth in IoTHub with that cert and then try to use that cert from teh client. That way we don't need the additional thumbprint parameters.. etc
I am only talking about Thumbprint Auth.. (it is clear for CA auth.. )

[the remark this answers is not in the document]
Comment (varunpuranik) [reply smoke/LeafDevice/LeafDevice.cs]: but it si fine either ways..

**Pull request merged**